### PR TITLE
Autofill capture an existing login new to Lockwise

### LINKED ISSUE #217: Save new credentials from Autofill
### Acceptance Criteria

Save credentials when the user enters credentials that we don't have in Lockbox. For now, this will not respect app locked state; that will be handled in a follow-up ticket #632 or saving values from multi-screen login experiences #633 

### Engineering notes 

Handle the [`onSaveRequest`](https://developer.android.com/reference/android/service/autofill/AutofillService.html#onSaveRequest(android.service.autofill.SaveRequest,%20android.service.autofill.SaveCallback)) callback.

### Todo
- [x] add SaveInfo to the FillResponse
- [x] `onSuccess` called in `LockboxAutofillService` 
- [x] add username and pw autofill ids as parcelables to the client state in `onFillRequest` -- [documentation](https://developer.android.com/reference/android/service/autofill/AutofillService.html#saving-when-data-is-split-in-multiple-screens)
- [x] Modify `ParsedStructure` to implement `findNodeById`
- [x] Use `getAutofillValues` + map to `ServerPassword`
- [x] Implement `DataStoreAction.Add(credential: ServerPassword)`
- [x] Handle `DataStoreAction.Add` in `Datastore.kt`

### PR DESCRIPTION
**Pull request by jhugman**
Fixes #217.

This PR introduces capturing of credentials from autofill contexts.

 * it is hidden behind a `FeatureFlags.AUTOFILL_CAPTURE`.
 * user can type a username and password into a screen, then once exiting the screen the offer to save the credential in Lockwise is given.
 * it is expected to work in both a well behaved browser (e.g. fenix) and native apps.

No duplicate checking is done, however, it may be possible that the autofill API tries its best not to do a save request. It is still possible, but not easy to force it to add a duplicate.

### CODE REVIEW

File: app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt [hunk 1]
```diff
@@ -121,27 +124,64 @@ class LockboxAutofillService(
...
+            .map { suffix ->
+                val domain = "https://${suffix.fullDomain}"
+                ServerPassword(
+                    id = "",
+                    hostname = domain,
+                    username = autofillItem.username,
+                    password = autofillItem.password ?: "",
+                    formSubmitURL = parsedStructure.webDomain ?: domain
+                )
+            }
```
Comment (eliserichards) [inline app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt]: Not sure how I feel about a half-completed `ServerPassword`

File: app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt [hunk 2]
```diff
@@ -121,27 +124,64 @@ class LockboxAutofillService(
...
+            .doOnComplete {
+                compositeDisposable.clear()
```
Comment (eliserichards) [inline app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt]: 👌

File: app/src/main/java/mozilla/lockbox/autofill/FillResponseBuilder.kt [hunk 3]
```diff
@@ -4,24 +4,34 @@ import android.annotation.TargetApi
...
+import mozilla.lockbox.support.Constant
 import mozilla.lockbox.support.PublicSuffixSupport
 import mozilla.lockbox.support.filter
 
 @TargetApi(Build.VERSION_CODES.O)
 @ExperimentalCoroutinesApi
 open class FillResponseBuilder(
-    internal val parsedStructure: ParsedStructure
+    internal val parsedStructure: ParsedStructure,
+    private val enableSave: Boolean = true
```
Comment (eliserichards) [inline app/src/main/java/mozilla/lockbox/autofill/FillResponseBuilder.kt]: This acts as the Feature Flag, correct? Will we change the signature once we remove the Feature Flags? I think I would prefer to just do a check [like we're doing for the delete/edit menu](https://github.com/mozilla-lockwise/lockwise-android/blob/0a272bed74f402f0022e784c9ab25d26ff94087d/app/src/main/java/mozilla/lockbox/presenter/ItemDetailPresenter.kt#L78)

File: app/src/main/java/mozilla/lockbox/autofill/FillResponseBuilder.kt [hunk 4]
```diff
@@ -31,24 +41,35 @@ open class FillResponseBuilder(
         presentation.setTextViewText(R.id.autofill_cta, authenticationText)
 
         val sender = IntentBuilder.getAuthIntentSender(context, this)
-        responseBuilder.setAuthentication(autofillIds(), sender, presentation)
+
+        responseBuilder.setAuthentication(parsedStructure.autofillIds, sender, presentation)
+        if (enableSave) {
+            responseBuilder.setSaveInfo(buildSaveInfo())
+        }
```
Comment (eliserichards) [inline app/src/main/java/mozilla/lockbox/autofill/FillResponseBuilder.kt]: This is duplicated here (`fun buildAuthenticationFillResponse`) and below in `fun buildFallbackFillResponse` and `fun buildFilteredFillResponse`. Consider making a single check on creation/initialization before these are even called? Alternatively, break this out into its own method.

File: app/src/main/java/mozilla/lockbox/autofill/ParsedStructureData.kt [hunk 5]
```diff
@@ -40,13 +41,37 @@ open class ParsedStructureData<Id>(
...
+        allIds.filterNotNull()
+            .toTypedArray()
+    }
+
+    val saveInfoMask: Int by lazy {
+        allIds.mapIndexed { index, id ->
+                id?.let { masks[index] } ?: 0
+            }
+            .reduce { acc, i ->
+                acc or i
```
Comment (eliserichards) [inline app/src/main/java/mozilla/lockbox/autofill/ParsedStructureData.kt]: Nit: better names needed

File: app/src/main/java/mozilla/lockbox/store/AccountStore.kt [hunk 6]
```diff
@@ -98,8 +98,9 @@ open class AccountStore(
     private val syncCredentials: Observable<Optional<SyncCredentials>> = ReplaySubject.createWithSize(1)
     open val profile: Observable<Optional<Profile>> = ReplaySubject.createWithSize(1)
 
-    private lateinit var webView: WebView
-    private lateinit var logDirectory: File
+    private val logDirectory: File
+        get() = context.getDir("webview", Context.MODE_PRIVATE)
```
Comment (eliserichards) [inline app/src/main/java/mozilla/lockbox/store/AccountStore.kt]: 👏 much better

File: app/src/main/java/mozilla/lockbox/store/AccountStore.kt [hunk 7]
```diff
@@ -312,7 +311,8 @@ open class AccountStore(
 
         this.generateNewFirefoxAccount()
 
-        webView.clearCache(true)
+        WebView(context).clearCache(true)
```
Comment (eliserichards) [inline app/src/main/java/mozilla/lockbox/store/AccountStore.kt]: Maybe add a comment here to explain why we're doing this? I think that might help clear up the ambiguity

File: app/src/main/java/mozilla/lockbox/store/DataStore.kt [hunk 8]
```diff
@@ -184,6 +186,44 @@ open class DataStore(
...
+            .addTo(compositeDisposable)
+    }
+
+    private fun backendEnsureLocked() =
+        backend.ensureLocked().asSingle(coroutineContext).toObservable()
+
+    private fun backendAdd(item: ServerPassword) =
+        backend.add(item).asSingle(coroutineContext).toObservable()
+
+    private fun backendEnsureUnlocked() =
```
Comment (eliserichards) [inline app/src/main/java/mozilla/lockbox/store/DataStore.kt]: Nit: have ensure locked and ensure unlocked next to each other

File: app/src/main/java/mozilla/lockbox/store/DataStore.kt [hunk 9]
```diff
@@ -184,6 +186,44 @@ open class DataStore(
         }
     }
 
+    private fun add(item: ServerPassword) {
+        val backend = this.backend
+        if (!backend.isLocked()) {
+            backend.add(item)
+                .asSingle(coroutineContext)
```
Comment (eliserichards) [inline app/src/main/java/mozilla/lockbox/store/DataStore.kt]: Use `backendAdd` here. Duplicate logic

File: app/src/test/java/mozilla/lockbox/support/AutofillTextValueBuilderTest.kt [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+import mozilla.lockbox.autofill.AutofillTextValueBuilder
+import mozilla.lockbox.autofill.ParsedStructure
+import mozilla.lockbox.autofill.ViewNodeNavigator
+import org.junit.Assert
+import org.junit.Test
+import org.mockito.Mock
+import org.mockito.Mockito.mock
+import org.mockito.Mockito.`when` as whenCalled
+
+class AutofillTextValueBuilderTest {
```
Comment (eliserichards) [inline app/src/test/java/mozilla/lockbox/support/AutofillTextValueBuilderTest.kt]: All the tests!!!!! 💯

**Comment by eliserichards** [comment]
autofill_capture_duolingo

File: app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt [hunk 11]
```diff
@@ -121,27 +121,88 @@ class LockboxAutofillService(
...
+        // "If login has an empty id field, then a GUID will be generated automatically."
+        val emptyId = ""
+
+        val pslSuffix =
+            parsedStructure.webDomain?.let { pslSupport.fromWebDomain(it) }
+            ?: pslSupport.fromPackageName(parsedStructure.packageName)
+
+        pslSuffix.take(1)
+            .map { suffix ->
+                val domain = "https://${suffix.fullDomain}"
```
Comment (mnoorenberghe) [inline app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt]: s/domain/origin/

File: app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt [hunk 12]
```diff
@@ -121,27 +121,88 @@ class LockboxAutofillService(
...
+                ServerPassword(
+                    id = emptyId,
+                    hostname = domain,
+                    username = capturedUsername,
+                    password = capturedPassword,
+                    formSubmitURL = parsedStructure.webDomain ?: domain
```
Comment (mnoorenberghe) [inline app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt]: Is `parsedStructure.webDomain` a domain or origin? `formSubmitURL` needs to be an origin.

**Pull request merged**